crates/wasmtime/src/component/types.rs

```diff
@@ -0,0 +1,1025 @@
+                1 + handle
+                    .types()
+                    .map(|ty| ty.flatten_count())
+                    .max()
+                    .unwrap_or(0)
+            }
+
+            Type::Option(handle) => 1 + handle.ty().flatten_count(),
+
+            Type::Expected(handle) => {
+                1 + handle
+                    .ok()
+                    .flatten_count()
+                    .max(handle.err().flatten_count())
+            }
+
+            Type::Flags(handle) => values::u32_count_for_flag_count(handle.names().len()),
+        }
+    }
+
+    fn desc(&self) -> &'static str {
+        match self {
+            Type::Unit => "unit",
+            Type::Bool => "bool",
+            Type::S8 => "s8",
+            Type::U8 => "u8",
+            Type::S16 => "s16",
+            Type::U16 => "u16",
+            Type::S32 => "s32",
+            Type::U32 => "u32",
+            Type::S64 => "s64",
+            Type::U64 => "u64",
+            Type::Float32 => "float32",
+            Type::Float64 => "float64",
+            Type::Char => "char",
+            Type::String => "string",
+            Type::List(_) => "list",
+            Type::Record(_) => "record",
+            Type::Tuple(_) => "tuple",
+            Type::Variant(_) => "variant",
+            Type::Enum(_) => "enum",
+            Type::Union(_) => "union",
+            Type::Option(_) => "option",
+            Type::Expected(_) => "expected",
+            Type::Flags(_) => "flags",
+        }
+    }
+
+    /// Deserialize a value of this type from core Wasm stack values.
+    pub(crate) fn lift<'a>(
+        &self,
+        store: &StoreOpaque,
+        options: &Options,
+        src: &mut impl Iterator<Item = &'a ValRaw>,
+    ) -> Result<Val> {
+        Ok(match self {
+            Type::Unit => Val::Unit,
+            Type::Bool => Val::Bool(bool::lift(store, options, next(src))?),
+            Type::S8 => Val::S8(i8::lift(store, options, next(src))?),
+            Type::U8 => Val::U8(u8::lift(store, options, next(src))?),
+            Type::S16 => Val::S16(i16::lift(store, options, next(src))?),
+            Type::U16 => Val::U16(u16::lift(store, options, next(src))?),
+            Type::S32 => Val::S32(i32::lift(store, options, next(src))?),
+            Type::U32 => Val::U32(u32::lift(store, options, next(src))?),
+            Type::S64 => Val::S64(i64::lift(store, options, next(src))?),
+            Type::U64 => Val::U64(u64::lift(store, options, next(src))?),
+            Type::Float32 => Val::Float32(u32::lift(store, options, next(src))?),
+            Type::Float64 => Val::Float64(u64::lift(store, options, next(src))?),
+            Type::Char => Val::Char(char::lift(store, options, next(src))?),
+            Type::String | Type::List(_) => {
+                // These won't fit in func::MAX_STACK_RESULTS as of this writing, so presumably we should never
+                // reach here
+                unreachable!()
```

Comment:
I think it would still be worth filling this in for other reasons though. For example all of the parameters to a host function are lifted which is where these could show up inline.

Eventually we'll want an equivalent to `crate::Linker::func_new` within `component::Linker` although it doesn't have to be added as part of that PR. When that's added this'll get exercised I believe.